python_modules/dagster/dagster/_core/definitions/asset_graph_differ.py

```diff
@@ -28,6 +41,47 @@ class AssetDefinitionChangeType(Enum):
     REMOVED = "REMOVED"
 
 
+class AssetDefinitionChangeData:
+    """Base interface for classes which
+    hold data about asset definition changes.
+    """
+
+    pass
+
+
+@whitelist_for_serdes
+@record
+class CodeVersionChangeData(AssetDefinitionChangeData):
+    base_code_version: Optional[str]
+    branch_code_version: Optional[str]
+
+
+@whitelist_for_serdes
+@record
+class DependenciesChangeData(AssetDefinitionChangeData):
+    base_dependencies: AbstractSet[
+        Annotated["AssetKey", ImportFrom("dagster._core.definitions.events")]
```

Comment:
Is this to avoid a circular import? If so, you can probably avoid it by importing from `dagster._core.definitions.asset_key`.

Reply:
> What do we get out of the more general form?

Are you asking about the difference between a single dataclass which contains all of the individual changes (to metadata, to tags etc) along with the list of change types, vs a dataclass which contains a list of tuples of (change type, change data)?

The main benefit with this approach I guess is that we can concretely associate change data with each change type

Reply:
Do we still need it to store the list of change types? Could that be derived from the other properties?

Reply:
Not something to rush into, because we'd need to figure out the UI, but could imagine it being useful to display the more detail diffs for branch deployments as well.